### PR TITLE
[Model] add model example CARE-GNN

### PR DESCRIPTION
**Pull request by kayzliu**
## Description

add model example CAmouflage-REsistant GNN from paper [Enhancing Graph Neural Network-based Fraud Detectors against Camouflaged Fraudsters](https://arxiv.org/pdf/2008.08692.pdf).
@zhjwy9343 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by zhjwy9343 — COMMENTED** [review]
Could you also change the example folder's readme to add this example on?

**Comment by kayzliu** [comment]
> Could you also change the example folder's readme to add this example on?

Sorry for the missing update, and it's modified in latest commit.

File: examples/pytorch/caregnn/model.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        sim = th.tanh(self.layers[0].MLP(feat))
+
+        # Forward of n layers of CARE-GNN
+        for i, layer in enumerate(self.layers):
+            feat = layer(graph, feat)
+
+        return feat, sim
+
+    def RLModule(self, graph, epoch, idx):
+        for i, layer in enumerate(self.layers):
```
Comment (zhjwy9343) [inline examples/pytorch/caregnn/model.py:146]: ditto

File: examples/pytorch/caregnn/model.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                    else:
+                        if layer.p[etype] + self.step_size <= 1:
+                            layer.p[etype] += self.step_size
+                        layer.f[etype].append(+1)
+                    layer.last_avg_dist[etype] = avg_dist
+
+                    # formula 7
+                    if epoch >= 9 and abs(sum(layer.f[etype][-10:])) <= 2:
+                        layer.cvg[etype] = True
+            # print(layer.p)
```
Comment (zhjwy9343) [inline examples/pytorch/caregnn/model.py:164]: Pls remove this line.

File: examples/pytorch/caregnn/model_sampling.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        # formula 8
+        hr = {}
+        for etype in g.canonical_etypes:
+            g.update_all(fn.copy_u('h', 'm'), fn.mean('m', 'hr'), etype=etype)
+            hr[etype] = g.dstdata['hr']
+            if self.activation is not None:
+                hr[etype] = self.activation(hr[etype])
+
+        # formula 9 using mean as inter-relation aggregator
+        h_homo = th.sum(th.stack(list(hr.values())) * th.Tensor(list(self.p.values())).view(-1, 1, 1).to(feat.device),
```
Comment (zhjwy9343) [inline examples/pytorch/caregnn/model_sampling.py:78]: Same comments as in the model.py line #69

File: examples/pytorch/caregnn/model_sampling.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        # formula 4
+        sim = th.tanh(self.layers[0].MLP(blocks[-1].dstdata['feature'].float()))
+
+        # Forward of n layers of CARE-GNN
+        for block, layer in zip(blocks, self.layers):
+            feat = layer(block, feat)
+        return feat, sim
+
+    def RLModule(self, graph, epoch, idx, dists):
+        for i, layer in enumerate(self.layers):
```
Comment (zhjwy9343) [inline examples/pytorch/caregnn/model_sampling.py:153]: Same comments as in model.py line #146

File: examples/pytorch/caregnn/model.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                sampled_edges = self._top_p_sampling(g[etype], self.p[etype])
+
+                # formula 8
+                g.send_and_recv(sampled_edges, fn.copy_u('h', 'm'), fn.mean('m', 'h_%s' % etype[1]), etype=etype)
+                hr[etype] = g.ndata['h_%s' % etype[1]]
+                if self.activation is not None:
+                    hr[etype] = self.activation(hr[etype])
+
+            # formula 9 using mean as inter-relation aggregator
+            h_homo = th.sum(th.stack(list(hr.values())) * th.Tensor(list(self.p.values())).view(-1, 1, 1).to(g.device),
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/caregnn/model.py]: Modified as per your requirement.

File: examples/pytorch/caregnn/model_sampling.py (hunk 3 above)
Comment (kayzliu) [reply examples/pytorch/caregnn/model_sampling.py]: Modified as per your requirement.

**Review by kayzliu — COMMENTED** [review]
Modified as per the requirements.

File: examples/pytorch/caregnn/main.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        # backward
+        optimizer.zero_grad()
+        tr_loss.backward()
+        optimizer.step()
+
+        # Print out performance
+        print("Epoch {}, Train: Recall: {:.4f} AUC: {:.4f} Loss: {:.4f} | Val: Recall: {:.4f} AUC: {:.4f} Loss: {:.4f}"
+              .format(epoch, tr_recall, tr_auc, tr_loss.item(), val_recall, val_auc, val_loss.item()))
+
+        model.RLModule(graph, epoch, rl_idx)
```
[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/caregnn/main.py:90]: Sure

File: examples/pytorch/caregnn/main_sampling.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    recall = 0
+    num_blocks = 0
+    for input_nodes, output_nodes, blocks in dataloader:
+        blocks = [b.to(device) for b in blocks]
+        feature = blocks[0].srcdata['feature'].float()
+        label = blocks[-1].dstdata['label'].squeeze().long()
+        logits_gnn, logits_sim = model(blocks, feature)
+
+        # compute loss
+        loss += loss_fn(logits_gnn, label) + args.sim_weight * loss_fn(logits_sim, label).item()
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply examples/pytorch/caregnn/main_sampling.py:22]: But so far you only has .item() at the second loss_fn, how about the first loss_fn?

[the remark this answers is not in the document]
Comment (kayzliu) [reply examples/pytorch/caregnn/main_sampling.py]: You're right. I have added it in the latest commit.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**